p2p/pex/pex_reactor.go

```diff
@@ -585,81 +592,65 @@ func (r *PEXReactor) crawlPeersRoutine() {
 func (r *PEXReactor) hasPotentialPeers() bool {
 	out, in, dial := r.Switch.NumPeers()
 
-	return out+in+dial > 0 && len(r.book.ListOfKnownAddresses()) > 0
+	return out+in+dial > 0 && r.book.Size() > 0
 }
 
-// crawlPeerInfo handles temporary data needed for the
-// network crawling performed during seed/crawler mode.
+// crawlPeerInfo handles temporary data needed for the network crawling
+// performed during seed/crawler mode.
 type crawlPeerInfo struct {
-	// The listening address of a potential peer we learned about
 	Addr *p2p.NetAddress
-
-	// The last time we attempt to reach this address
-	LastAttempt time.Time
-
-	// The last time we successfully reached this address
-	LastSuccess time.Time
+	// The last time we crawled the peer or attempted to do so.
+	LastCrawled time.Time
 }
 
-// oldestFirst implements sort.Interface for []crawlPeerInfo
-// based on the LastAttempt field.
-type oldestFirst []crawlPeerInfo
-
-func (of oldestFirst) Len() int           { return len(of) }
-func (of oldestFirst) Swap(i, j int)      { of[i], of[j] = of[j], of[i] }
-func (of oldestFirst) Less(i, j int) bool { return of[i].LastAttempt.Before(of[j].LastAttempt) }
-
-// getPeersToCrawl returns addresses of potential peers that we wish to validate.
-// NOTE: The status information is ordered as described above.
-func (r *PEXReactor) getPeersToCrawl() []crawlPeerInfo {
-	var of oldestFirst
+// CrawlPeers will crawl the network looking for new peer addresses.
+// Exposed for testing.
+func (r *PEXReactor) CrawlPeers(addrs []*p2p.NetAddress) {
+	now := time.Now()
 
-	// TODO: be more selective
-	addrs := r.book.ListOfKnownAddresses()
 	for _, addr := range addrs {
-		if len(addr.ID()) == 0 {
-			continue // dont use peers without id
+		peerInfo, ok := r.crawlPeerInfos[addr.ID]
+		if !ok {
+			peerInfo = &crawlPeerInfo{
+				Addr:        addr,
+				LastCrawled: now.Add(-defaultCrawlPeerInterval).Add(-1 * time.Second),
```

Comment:
This looks a bit random to me. Suggestion: Maybe add a comment why we subtract a second here (additionally to `defaultCrawlPeerInterval` which is 2 mins).

Reply:
To ensure `now.Sub(peerInfo.LastCrawled) < defaultCrawlPeerInterval` is false the first time we add the peer